casper/src/main/rholang/MakePoS.rho

```diff
@@ -183,6 +186,48 @@ in {
           }
         } |
       
+        //TODO: Fill in this contract (RHOL-462)
+        contract this(@"distribute", @purse) = {
+          Nil
+        } |
+      
+        paymentStore!((Nil, Nil, Nil, Nil)) |
+      
+        //Make it easy to query the last payment so that validators
+        //can check how much phlo there is for the deploy.
+        contract this(@"lastPayment", return) = {
+          for(@payment <- paymentStore) {
+            paymentStore!(payment) |
+            return!(payment)
+          }
+        } |
+      
+        //This is the contract that deploy payment code ultimately needs
+        //to call in order to pay for the deploy.
+        contract this(@"pay", @purse, return) = {
+          new payPurseCh, p(`rho:deploy:params`), paramsCh in {
+            clonePurse!(purse, *payPurseCh) |
+            p!(*paramsCh) |
+            for(
+              @cloneResult <- payPurseCh;
+              @codeHash, @phloRate, @userId, @timestamp <- paramsCh
+            ) {
+              match cloneResult {
+                ("Right", (payPurse, payAmount)) => {
+                  this!("distribute", payPurse) | //distribute rev paid to validators
+                  //update paymentStore
+                  for(_ <- paymentStore){ paymentStore!((codeHash, userId, timestamp, phloRate * payAmount)) } |
+                  return!(phloRate * payAmount) //return amount of phlo purchased to user (in case they care)
+                }
+                ("Left", message) => {
+                  for(_ <- paymentStore){ paymentStore!((codeHash, userId, timestamp, 0)) } |
+                  return!(0)
+                }
+              }
+            }
+          }
+        } |
```

Comment:
If you didn't store the deploy params on the `paymentStore` channel this contract could be reduced to:
```
clonePurse!(purse, *payPurseCh) | 
for(@cloneResult <- payPurseCh) { 
  match cloneResult {
    ("Right", (payPurse, payAmount)) => {
      this!("distribute", payPurse) | …
```
am I right?

Reply:
Yes. But I think storing the last payment is useful. The new "short leash deploy" workflow I am imagining is validators run the "payment code", then inspect what was paid via the `lastPayment` method and proceed to run the actual deploy with the appropriate amount of phlo. It could be  that the user's payment code did not call the pay API, in which case `lastPayment` will not have changed; so I include the extra information in order to identify the correct deploy has paid.

Reply:
Approach I took in my [PR](https://github.com/rchain/rchain/pull/1757/files#diff-8172c898e2d53ee886672b92b2e248c9R182) is that the deployment of payment code returns phlos used.